src/main/java/fr/sncf/osrd/infra_state/SwitchPost.java

```diff
@@ -0,0 +1,204 @@
+package fr.sncf.osrd.infra_state;
+
+import fr.sncf.osrd.SuccessionTable;
+import fr.sncf.osrd.infra.TVDSection;
+import fr.sncf.osrd.infra.routegraph.Route;
+import fr.sncf.osrd.train.Train;
+import fr.sncf.osrd.train.TrainStatus;
+import fr.sncf.osrd.simulation.Simulation;
+import fr.sncf.osrd.simulation.SimulationError;
+import fr.sncf.osrd.infra.Infra;
+
+import java.util.List;
+import java.util.HashMap;
+import java.util.HashSet;
+
+/**
+ * The SwitchPost is an improved waiting queue designed to handle the routes
+ * reservation requests
+ */
+public class SwitchPost {
+
+    private HashMap<String, SuccessionTable> tables;
+    private HashMap<String, Integer> currentIndex;
+    private HashMap<String, HashMap<String, Integer>> occurences;
+    private HashMap<String, HashSet<Request>> waitingList;
+    private HashMap<String, String> lastRequestedRoute;
+    private HashMap<String, String> currentTrainAllowed;
+
+    public SwitchPost() {
+        tables = null;
+    }
+
+    /**
+     * Initialize the SwitchPost with an infrastructure and trains successions
+     * tables
+     * 
+     * @param infra      the infrastructure
+     * @param initTables the initial trains successions tables
+     */
+    public void init(Infra infra, List<SuccessionTable> initTables) {
+        // build succession tables from initTables
+        tables = new HashMap<String, SuccessionTable>();
+        currentIndex = new HashMap<String, Integer>();
+        occurences = new HashMap<String, HashMap<String, Integer>>();
```

Comment:
Most of the parameters between `<>` are not needed here. If you use intellij they should appear in gray.